sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/GroupIntoBatches.java

```diff
@@ -105,23 +108,83 @@ public long getBatchSize() {
   }
 
   /**
-   * Set a time limit (in processing time) on how long an incomplete batch of elements is allowed to
-   * be buffered. Once a batch is flushed to output, the timer is reset.
+   * Sets a time limit (in processing time) on how long an incomplete batch of elements is allowed
+   * to be buffered. Once a batch is flushed to output, the timer is reset.
    */
   public GroupIntoBatches<K, InputT> withMaxBufferingDuration(Duration duration) {
     checkArgument(
         duration.isLongerThan(Duration.ZERO), "max buffering duration should be a positive value");
     return new GroupIntoBatches<>(batchSize, duration);
   }
 
+  /**
+   * Outputs batched elements associated with sharded input keys. The sharding is determined by the
+   * runner to balance the load during the execution time. By default, apply no sharding so each key
+   * has one shard.
+   */
+  @Experimental
+  public WithShardedKey withShardedKey() {
+    return new WithShardedKey();
+  }
+
+  public class WithShardedKey
+      extends PTransform<
+          PCollection<KV<K, InputT>>, PCollection<KV<ShardedKey<K>, Iterable<InputT>>>> {
+
+    /** Returns the size of the batch. */
+    public long getBatchSize() {
+      return batchSize;
+    }
+
+    /** Returns the size of the batch. */
+    @Nullable
+    public Duration getMaxBufferingDuration() {
+      return maxBufferingDuration;
+    }
+
+    @Override
+    public PCollection<KV<ShardedKey<K>, Iterable<InputT>>> expand(
+        PCollection<KV<K, InputT>> input) {
+      Duration allowedLateness = input.getWindowingStrategy().getAllowedLateness();
+
+      checkArgument(
+          input.getCoder() instanceof KvCoder,
+          "coder specified in the input PCollection is not a KvCoder");
+      KvCoder<K, InputT> inputCoder = (KvCoder<K, InputT>) input.getCoder();
+      Coder<K> keyCoder = (Coder<K>) inputCoder.getCoderArguments().get(0);
+      Coder<InputT> valueCoder = (Coder<InputT>) inputCoder.getCoderArguments().get(1);
+
+      return input
+          .apply(
+              MapElements.via(
+                  new SimpleFunction<KV<K, InputT>, KV<ShardedKey<K>, InputT>>() {
+                    @Override
+                    public KV<ShardedKey<K>, InputT> apply(KV<K, InputT> input) {
+                      // By default every input key has only one shard.
+                      return KV.of(
+                          ShardedKey.of(input.getKey(), DEFAULT_SHARD_ID), input.getValue());
```

Comment:
A single subshard by default will make this virtually unusable for runners that don't implement the optimization (including batch Dataflow). Instead use something like the thread id here, or at the very lease initialize DEFAULT_SHARD_ID to be different for each worker and add a small nonce. We could alternatively take a hint as to the number of subshards that would be nice (but that has its own downsides).

Reply:
Good point! I changed the default implementation to use thread id as shard id.